### PR TITLE
Add a fetch test for upload streaming

### PR DESCRIPTION
**Pull request by yutakahirano**
Adding a test for uploading a request made from a ReadableStream.

### CODE REVIEW

**Comment by yutakahirano** [comment]
@annevk: This is for https://github.com/whatwg/fetch/pull/425

**Comment by annevk** [comment]
This looks okay to me. Though ideally we also have some tests around redirect handling and different kinds of methods. For a new feature that will need tests written anyway as part of the implementation this might be good enough though.

@foolip @domenic thoughts?

**Comment by foolip** [comment]
Yeah, testing non-trivials things blindly is very hard, and a good reason to not write all tests up front. Keeping track somewhere of what testing coverage is missing would be nice of course, but only if someone is then using that to guide testing priorities.

**Comment by domenic** [comment]
Wait, we're allowing strings to be enqueued? I didn't get that from https://github.com/whatwg/fetch/pull/425 .

**Comment by annevk** [comment]
That is a good point and needs to be fixed in the test.

Where do we currently place restrictions on what can be on a `ReadableStream` object used in Fetch? It seems like we don't...

**Comment by yutakahirano** [comment]
> Wait, we're allowing strings to be enqueued? I didn't get that from whatwg/fetch#425 .

Sorry that was my mistake. Fixed.

> Where do we currently place restrictions on what can be on a ReadableStream object used in Fetch? It seems like we don't...

At least, Request.text(), etc. expect that as they are using https://fetch.spec.whatwg.org/#concept-read-all-bytes-from-readablestream.

**Comment by annevk** [comment]
We should also make the network layer expect that though. And whatever ends up reading a response.

**Comment by yutakahirano** [comment]
At https://fetch.spec.whatwg.org/#read-a-request?

**Comment by annevk** [comment]
Yeah I suppose. And we probably need something similar for responses then and make sure all specifications use it… 😟

**Comment by guest271314** [comment]
@annevk Following https://www.w3.org/2015/05/Web_Platform_Testing_Latest_News.pdf, is the current test able to be run in browser at http://w3c-test.org/fetch/api/ ?

**Comment by annevk** [comment]
I'm not sure what you're asking for.

**Comment by guest271314** [comment]
@annevk Tried to run the current test in browser. Checked all fetch tests, browser crashed at a worker test. Just want to run the current test. How to do this at *nix, chromium 52?

**Comment by annevk** [comment]
What current test? This is a pull request for a test change. It hasn't landed...

**Comment by guest271314** [comment]
@annevk Ok. Was not certain about that. Not well-versed in the terminology and mechanics of git, pull requests. Is it not currently possible to try this now? Is a "pull request" code written that is tested before being published at large?

**Comment by annevk** [comment]
You can locally check out the code and run it https://github.com/w3c/web-platform-tests/blob/master/README.md. I believe for some PRs there's also a URL you can visit, but not for this PR it seems (might depend on where the branch is situated).

**Comment by guest271314** [comment]
@annevk Is the above referenced test in that repository? Do not need to run all of the tests in the repo, just the newest one, or only the `fetch` tests. Is it necessary to download the entire repo? And install the python server? Was there changes made to the way `fetch` `body` functions in order to pass the test?

**Comment by annevk** [comment]
The test of this PR is in https://github.com/yutakahirano/web-platform-tests/tree/upload-streaming, a branch in a fork of this repository. If you just want to know whether any browser passes the test, I can tell you that none do, since as I told you elsewhere, this is not implemented yet.

**Comment by guest271314** [comment]
@annevk Actually seeking more knowledge than just to know if currently implemented; also how implemented at tests? Or, that is, how is the test run if not implemented? The author(s) composed code which is being tested? A custom browser build implementing the changes necessary to pass the test? Gracious for your patience here. Again, just attempting to learn. If this exchange is not helpful for the topic, here, can stop now and try to gather pieces of information to possibly answer questions, at large.

**Comment by guest271314** [comment]
@annevk Am attempting to gain more knowledge of the process. Your feedback has been helpful.

**Comment by guest271314** [comment]
@annevk  Is the appropriate venue to ask questions about this implementation https://github.com/whatwg/fetch/pull/425 ?

**Comment by annevk** [comment]
@yutakahirano perhaps you can also add a test that expects strings and numbers and such to fail.

**Comment by annevk** [comment]
Thanks, @domenic could you give it another review?

File: fetch/api/basic/request-upload.js [hunk 1]
```diff
@@ -18,6 +18,20 @@ function testUpload(desc, url, method, body, expectedBody) {
...
 }
 
+function testUploadFailure(desc, url, method, body) {
+  const requestInit = {"method": method};
+  promise_test(test => {
+    if (typeof body === "function")
+      body = body();
+    if (body)
+      requestInit["body"] = body;
+    return fetch(url, requestInit).then(() => {
```
Comment (domenic) [inline fetch/api/basic/request-upload.js]: This should use `return promise_rejects(new TypeError(), fetch(url, requestInit))`

File: fetch/api/basic/request-upload.js [hunk 2]
```diff
@@ -31,5 +45,22 @@ testUpload("Fetch with POST with Float32Array body", url, "POST", new Float32Arr
...
+    controller.close();
+  }}));
+testUploadFailure("Fetch with POST with ReadableStream containing null", url, "POST", new ReadableStream({start: controller => {
+    controller.enqueue(null);
+    controller.close();
+  }}));
+testUploadFailure("Fetch with POST with ReadableStream containing number", url, "POST", new ReadableStream({start: controller => {
+    controller.enqueue(99);
+    controller.close();
+  }}));
```
Comment (domenic) [inline fetch/api/basic/request-upload.js]: We should also test ArrayBuffer and Blob are not allowed

**Review by domenic — CHANGES_REQUESTED** [review]
Some minor test nits. Additionally I have something I am confused about; forgive me if we have discussed it before.

These tests seem to imply that fetch() will read the whole request body stream before fulfilling or rejecting the fetch promise. But, doesn't that hurt our ability to do full-duplex communication with streams? I thought the idea was that you could pass a request body stream, and get back a response body stream, and use them both at the same time. If we wait for the request body stream to be fully read before vending the response body stream, this is not possible, right?

**Comment by annevk** [comment]
I agree with your assessment of how things should be with regards to full-duplex communication. These tests are simply not testing that scenario, or am I missing something?

**Comment by annevk** [comment]
I would expect that the default server setup is such that it doesn't emit a response before getting the full request body. It would be good to have tests for early transmitted responses though.

**Comment by domenic** [comment]
I see. So is the intent of the spec that there's a race then? If the fetch gets terminated before the promise settles, then the promise is rejected (as in this case); if the fetch gets terminated after the response comes back, the promise is still fulfilled?

**Comment by annevk** [comment]
Yeah, the fetch gets terminated so you might not get the whole response body. We should really figure out a way to write full duplex tests since I suspect it is all kinds of broken.

**Comment by domenic** [comment]
OK, if that's intended then the tests are good, just the nits in my review remaining.

File: fetch/api/basic/request-upload.js (hunk 2 above)
Comment (yutakahirano) [inline fetch/api/basic/request-upload.js]: Done.

File: fetch/api/basic/request-upload.js (hunk 1 above)
Comment (yutakahirano) [inline fetch/api/basic/request-upload.js]: Done.

**Review by domenic — APPROVED** [review]

**Pull request merged**